### PR TITLE
[SPARK-24027][SQL] Support MapType with StringType for keys as the root type by from_json

### PR DESCRIPTION
**Pull request by MaxGekk**
## What changes were proposed in this pull request?

Currently, the from_json function support StructType or ArrayType as the root type. The PR allows to specify MapType(StringType, DataType) as the root type additionally to mentioned types. For example:

```scala
import org.apache.spark.sql.types._
val schema = MapType(StringType, IntegerType)
val in = Seq("""{"a": 1, "b": 2, "c": 3}""").toDS()
in.select(from_json($"value", schema, Map[String, String]())).collect()
```
```
res1: Array[org.apache.spark.sql.Row] = Array([Map(a -> 1, b -> 2, c -> 3)])
```

## How was this patch tested?

It was checked by new tests for the map type with integer type and struct type as value types. Also roundtrip tests like from_json(to_json) and to_json(from_json) for MapType are added.

### CODE REVIEW

**Comment by gatorsmile** [comment]
add to whitelist

**Comment by MaxGekk** [comment]
jenkins, retest this, please

**Comment by MaxGekk** [comment]
@gatorsmile Please, review the PR

**Comment by MaxGekk** [comment]
jenkins, retest this, please

**Comment by MaxGekk** [comment]
jenkins, retest this, please

**Comment by MaxGekk** [comment]
jenkins, retest this, please

**Comment by MaxGekk** [comment]
@gatorsmile Kindly ask you to look at it again.

**Comment by MaxGekk** [comment]
@marmbrus May I ask you to look at the PR.

File: sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala [hunk 1]
```diff
@@ -326,4 +326,61 @@ class JsonFunctionsSuite extends QueryTest with SharedSQLContext {
...
+        |  "valueContainsNull" : true
+        |}
+      """.stripMargin
+    val out = in.select(from_json($"value", schema, Map[String, String]()))
+
+    assert(out.columns.head == "entries")
+    checkAnswer(out, Row(Map("a" -> 1, "b" -> 2, "c" -> 3)))
+  }
+
+  test("SPARK-24027: from_json - map<string, struct>") {
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala]: Any negative test case? For example, when the key of MapType is not StringType?

Comment (MaxGekk) [reply sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala]: I added a negative test. Please, take a look at the PR again.

**Comment by gatorsmile** [comment]
retest this please

**Comment by gatorsmile** [comment]
Thanks! Merged to master.

**Pull request closed without merging**